subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheKeyIntegrationTest.groovy

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache
+
+
+import org.gradle.configurationcache.isolated.IsolatedProjectsFixture
+import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
+import org.gradle.test.fixtures.file.TestFile
+import spock.lang.IgnoreIf
+import spock.lang.Issue
+
+class ConfigurationCacheKeyIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
+
+    def configurationCache = newConfigurationCacheFixture()
+
+    def "offline flag is part of the cache key"() {
+        when:
+        configurationCacheRun "help"
+        then:
+        configurationCache.assertStateStored()
+        countStoredEntries() == 1
+
+        when:
+        configurationCacheRun "help", "--offline"
+        then:
+        configurationCache.assertStateStored()
+        countStoredEntries() == 2
+
+        // Now repeat invocations in different order to make sure both entries can be reused
+        when:
+        configurationCacheRun "help"
+        then:
+        configurationCache.assertStateLoaded()
+        countStoredEntries() == 2
+
+        when:
+        configurationCacheRun "help", "--offline"
+        then:
+        configurationCache.assertStateLoaded()
+        countStoredEntries() == 2
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/26049")
+    @IgnoreIf({ GradleContextualExecuter.isolatedProjects }) // Isolated Projects option is explicitly controlled by the test
+    def "isolated projects flag is part of the cache key"() {
+        def isolatedProjects = new IsolatedProjectsFixture(this)
+
+        when:
+        configurationCacheRun "help"
+        then:
+        configurationCache.assertStateStored()
+        countStoredEntries() == 1
```

Comment:
❓ Why?

Reply:
Initially, I wanted to test that the different runs actually use the exact same CC key. However, I found that we do not expose those keys neither in the logs nor in build operations.

Checking the number of CC entry directories is the closest reasonable approximation I found to checking that.

If your question is targeted towards the very first check with `== 1`, it's just for unification, so that we check after every task invocation in the test.

Reply:
> Initially, I wanted to test that the different runs actually use the exact same CC key

In that case I'd add a test to `ConfigurationCacheKeyTest` similar to https://github.com/gradle/gradle/blob/4be5615ce33258a8315ff7ef176df00011da83bd/subprojects/configuration-cache/src/test/kotlin/org/gradle/configurationcache/ConfigurationCacheKeyTest.kt#L91

For integration tests, the `assertStateStored` / `assertStateLoaded` should be enough. I wouldn't make this test depend on the cache layout on disk.

Reply:
It makes sense, I'll fix this.

I saw the unit test, but decided against it, because there we would have to start duplicating some logic of computing the `BuildModelParameters` from `StartParameterInternal`. We can do this, but only partially. I think it could be too easy for the production code and the unit test code to desynchronize.

Having said that, having both the integration test and the unit test should be good enough.

Reply:
Added a unit test, and removed the CC layout checks in the integration test